### ISSUE TITLE
Different Prediction for adversarial images when saved to disk 

### ISSUE THREAD

**Issue by anishmadan23** [bethgelab/foolbox#264]
I am facing a problem while computing predictions for adversaries after saving them to disk. I am using a **VGG16 model in PyTorch**. When foolbox generates an adversary and I try to get prediction details, it works fine, and the output is a different class with some prediction probability( note that the criterion used to generate is **Misclassification** and the attack used is **RandomStartProjectedGradientDescentAttack**). But now when I save this adversarial image and load it again to get prediction details, the class is sometimes different to the adversarial class predicted earlier( and is mostly the original/ground truth class) and the prediction probability is different everytime. 
`if torch.cuda.is_available():
    device = torch.device('cuda')
else:
    device = torch.device('cpu')
vgg = models.vgg16(pretrained=True).eval()
vgg = vgg.to(device)
vgg.eval()`

`
class craftAdversaries():

    def __init__(self,model, device, attack):
        self.model = model
        self.device = device
        self.attack = attack

    def getPredictionInfo(self,img,probOfClass=None):
        output = self.model(img)
        _,pred = torch.max(output.data,1)
        #     print(adv_img.data-img.data)
        op_probs = F.softmax(output, dim=1)                 #get probability distribution over classes
        pred_prob =  ((torch.max(op_probs.data, 1)[0][0]) * 100, 4)      #find probability (confidence) of a predicted class
        
        if probOfClass is not None:
            prob_class_of_interest = op_probs[:,probOfClass]
            return output,pred,op_probs,pred_prob,int(prob_class_of_interest[0].cpu())
        
        return output,pred,op_probs,pred_prob
    
    
    def attackImage(self,cur_img,img_name,label, cur_img_path, save_dir):
        
        mean = np.array([0.485, 0.456, 0.406]).reshape((3, 1, 1))
        std = np.array([0.229, 0.224, 0.225]).reshape((3, 1, 1))
        
        ##### initialise PyTorch Foolbox model #########
        fmodel = foolbox.models.PyTorchModel(
            self.model, bounds=(0, 1), num_classes=1000, preprocessing=(mean, std))

        normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                             std=[0.229, 0.224, 0.225])
        preprocess = transforms.Compose([transforms.Resize((256,256)),
                   transforms.CenterCrop(224),
                   transforms.ToTensor(),
                   normalize])
        
        ########## adversary has different preprocessing steps since we save the adv image generated and is already 224x224
        preprocess_adv = transforms.Compose([
                   transforms.ToTensor(),
                   normalize])
        
        
        ######## preprocessing steps to feed image to foolbox model ######3
        image = Image.open(cur_img_path).convert('RGB')
        img_for_tsor = image.copy()
        
        resize_transform = transforms.Resize((256,256))
        image = resize_transform(image)

        crop_transform = transforms.CenterCrop(224)
        image = crop_transform(image)

        image = np.asarray(image, dtype=np.float32)
        image = image[:, :, :3]
        print(img_name,image.shape)
        image = image.transpose((2,0,1))
        print(img_name,image.shape)
        image = image / 255.
        
        predicted_label = np.argmax(fmodel.predictions(image))

    ################ apply attack on image #####################################
        
        
        
        img_tsor = preprocess(img_for_tsor)
        img_tsor = img_tsor.to(self.device)
        img_tsor.unsqueeze_(0)
        _,unp_pred,_,unp_pred_prob = self.getPredictionInfo(img_tsor)
        print('Unperturbed Pred Prob',unp_pred,unp_pred_prob)
        
        
        print(label,predicted_label)
        
        if label==predicted_label:       # helps check if foolbox model is predicting correctly
             
            try:    
                save_true_dir_name = str('./test_')+str(label)+str('_true/')

                if not os.path.exists(save_true_dir_name):
                    os.makedirs(save_true_dir_name)

                if not os.path.exists(save_dir):
                    os.makedirs(save_dir)

                save_true_img_path = os.path.join(save_true_dir_name,img_name)

    #                 print(img_path)
    #                 print(save_true_img_path)

                ####### do attack ###########
                criterion = Misclassification()
                attack = self.attack(model = fmodel,distance=Linfinity,criterion=criterion)
                adversarial = attack(image, label)

                print('Adv shape',adversarial.shape,type(adversarial))

                ## adversarial is ndarray, so convert to tensor and predict
                adv_tsor = torch.from_numpy(adversarial).to(device)
                adv_tsor = normalize(adv_tsor)
                adv_tsor.unsqueeze_(0)
                _,adv_pred,_,adv_pred_prob,origProb_in_adv = self.getPredictionInfo(adv_tsor,label)
                print('Prediction from adversary', int(adv_pred.cpu()),float(adv_pred_prob[0].cpu()))

                adv_label = np.argmax(fmodel.predictions(adversarial))       # foolbox prediction 

                print('AdvInfoFromFoolbox',adv_label)

                #####3 saving adversary image#####
                save_name = str(img_name[:img_name.find('.')])+str('_')+str(int(round(unp_pred_prob[0])))+str('_')+str(int(adv_pred.cpu()))+str('_')+str(int(round(adv_pred_prob[0])))+str('_')+str(origProb_in_adv)+str('.png')
                save_path = os.path.join(save_dir,save_name)

                imsave(save_path,adversarial.transpose((1,2,0)))         # saves in 0-255 format

                ######## load image and try to predict yourself. This part predicts differently( different label sometimes and different prediction probability)  ############
                adv_my_img=Image.open(save_path).convert('RGB')
                adv_img_tsor = preprocess_adv(adv_my_img).cuda()
                adv_img_tsor.unsqueeze_(0)
                _,my_adv_pred,_,my_adv_pred_prob = self.getPredictionInfo(adv_img_tsor)
                print()
                print(save_name,'My_pred',my_adv_pred,my_adv_pred_prob)

                shutil.copy(img_path,save_true_img_path)         # copy only if adversary is found for that image else go to exception
            except:
                print('No adversary found!')
        else:
        
            print('Unperturbed pred not matching to ground truth label')
            print(img_name)`

`
np.random.seed(0)

img_dirs = ['test_768']
makeAdversary = craftAdversaries(vgg,device,RandomStartProjectedGradientDescentAttack)
for img_dir in img_dirs:

    adv_dir_name = str('./')+str(img_dir)+str('_pgd')

    imgs_ = os.listdir(img_dir)
    
    for img_ in imgs_:
        img_path = os.path.join(img_dir,img_)
        img = Image.open(img_path).convert('RGB')
        label = img_dir.split('_')[1]
        makeAdversary.attackImage(img,img_,int(label),img_path,adv_dir_name)`

The main lines where the issue is highlighted are:
`print('Prediction from adversary', int(adv_pred.cpu()),float(adv_pred_prob[0].cpu()))`
`print(save_name,'My_pred',my_adv_pred,my_adv_pred_prob)`

 All the images for whose adversary I want are in the folder 'test_768'( It actually contains images from the 768 class of ImageNet).

These two differ in their results and hence my issue. Any insight into this issue is much appreciated.
PS: Sorry for the bad formatting.

**Comment by anishmadan23** [comment]
Thanks for the reply. I have changed from saving to jpg or png to saving as .mat files and it works now.
